contracts/Curation.sol

```diff
@@ -0,0 +1,415 @@
+        IRegistry _registry,
+        IStaking _staking,
+        IVoting _voting,
+        uint256 _minDeposit,
+        uint64 _applyStageLen,
+        uint256 _dispensationPct
+    )
+        onlyInit
+        external
+    {
+        initialized();
+
+        require(isContract(_registry));
+        require(isContract(_staking));
+
+        registry = _registry;
+        staking = _staking;
+        _setVotingApp(_voting);
+
+        _setMinDeposit(_minDeposit);
+        _setApplyStageLen(_applyStageLen);
+        _setDispensationPct(_dispensationPct);
+    }
+
+    function newApplication(bytes data, uint256 lockId) isInitialized public returns (bytes32 entryId) {
+        entryId = keccak256(data);
+
+        require(data.length != 0);
+        // check data doesn't have an ongoing application
+        require(!applicationExists(entryId));
+        // check data doesn't exist in Registry
+        require(!registry.exists(entryId));
+
+        // check locked tokens
+        uint256 amount = _checkLock(msg.sender, lockId, MAX_UINT64);
+
+        applications[entryId] = Application({
+            applicant: msg.sender,
+            date: getTimestamp(),
+            registered: false,
+            data: data,
+            amount: amount,
+            lockId: lockId
+        });
+
+        // register used lock
+        NewApplication(entryId, msg.sender);
+    }
+
+    function challengeApplication(bytes32 entryId, uint256 lockId) isInitialized public returns(bytes32) {
+        // check application doesn't have an ongoing challenge
+        require(!challengeExists(entryId));
+        // check locked tokens
+        uint256 amount = _checkLock(msg.sender, lockId, getTimestamp().add(applyStageLen));
+
+        // touch-and-remove case
+        Application memory application = applications[entryId];
+        if (application.amount < minDeposit) {
+            registry.remove(entryId);
+            staking.unlock(application.applicant, application.lockId);
+            staking.unlock(msg.sender, lockId);
+            return 0;
+        }
+
+        // create vote
+        // TODO: metadata
+        // script to call `resolveChallenge(entryId)`
+        uint256 scriptLength = 64; // 4 (spec) + 20 (address) + 4 (calldataLength) + 4 (signature) + 32 (input)
+        bytes4 spec = bytes4(0x01);
+        address target = address(this);
+        bytes memory targetBytes = new bytes(32);
+        bytes4 calldataLength = bytes4(0x24); // 4 + 32
+        bytes4 signature = this.resolveChallenge.selector;
+        bytes memory executionScript = new bytes(scriptLength);
+        // concatenate spec + address(this) + calldataLength + calldata
+        // TODO: should we put this somewhere in aragonOS to be able ti reuse? (if it's not already there)
+        assembly {
+            mstore(add(executionScript, 0x20), spec)
+            mstore(add(targetBytes, 0x20), target)
+            mstore(add(executionScript, 0x24), mload(add(targetBytes, 0x2C)))
```

Comment:
Why not store the target address directly here?

Reply:
It would be better to just `mload(target)`, shift the required bits and mstore that. Otherwise we are reading from unknown memory and that's a little scary.

Reply:
I don't get why we are reading from unknown memory, but my assembly understanding is limited yet. Anyway, you are right, this way is too messy, I did it as you say, just shifting the address, way better. Thanks!